test/ILLink.RoslynAnalyzer.Tests/TestChecker.cs

```diff
@@ -16,51 +17,147 @@
 
 namespace ILLink.RoslynAnalyzer.Tests
 {
-	internal class TestChecker
+	internal class TestChecker : CSharpSyntaxWalker
 	{
-		private readonly CompilationWithAnalyzers Compilation;
+		private readonly CSharpSyntaxTree _tree;
+		private readonly SemanticModel _semanticModel;
+		private readonly IReadOnlyList<Diagnostic> _diagnostics;
+		private readonly List<Diagnostic> _unmatched;
+		private readonly List<(AttributeSyntax Attribute, string Message)> _missing;
+
+		public TestChecker (
+			CSharpSyntaxTree tree,
+			SemanticModel semanticModel,
+			ImmutableArray<Diagnostic> diagnostics)
+		{
+			_tree = tree;
+			_semanticModel = semanticModel;
+			_diagnostics = diagnostics
+				// Filter down to diagnostics which originate from this tree
+				.Where (d => d.Location.SourceTree == tree).ToList ();
+
+			// Filled in later
+			_unmatched = new List<Diagnostic> ();
+			_missing = new List<(AttributeSyntax Attribute, string Message)> ();
+		}
+
+		public void Check ()
+		{
+			_unmatched.Clear ();
+			_unmatched.AddRange (_diagnostics);
+			_missing.Clear ();
+
+			Visit (_tree.GetRoot ());
+
+			string message = "";
+			if (_missing.Any ()) {
+				var missingLines = string.Join (
+					Environment.NewLine,
+					_missing.Select (md => $"({md.Attribute.GetLocation ().GetLineSpan ()}) {md.Message}"));
+				message += $@"Missing diagnostics:{Environment.NewLine}{missingLines}
+";
+			}
+			if (_unmatched.Any ()) {
 
-		private readonly SemanticModel SemanticModel;
+				message += $"Found unmatched diagnostics:{Environment.NewLine}{string.Join (Environment.NewLine, _unmatched)}";
```

Comment:
"Unexpected diagnostic was generated" (similar to the above - unmatched to what?)

Personally I would go one step further and use "warning" instead of "diagnostic", but it's minor (mainly to match the test attributes ExpectedWarning and such)